_posts/2021-11-18-keras-updates/keras-updates.Rmd

```diff
@@ -0,0 +1,159 @@
+
+The mention of text vectorization leads to a second advantage. Unlike, say, a random distortion, vectorization is not something that, once done, may be forgotten about. We don't want to lose the original information, namely, the words. The same happens, for numerical data, with normalization. We need to keep the summary statistics. This means there are two types of pre-processing layers: stateless and stateful ones. The former are part of the training process; the latter are called in advance.
+
+Stateless layers, on the other hand, can appear in two places in the training workflow: as part of the `tfdatasets` pipeline, or as part of the model.
+
+This is, schematically, how the former would look.
+
+```{r}
+library(tfdatasets)
+dataset <- ... # define dataset
+dataset <- dataset %>%
+  dataset_map(function(x, y) list(preprocessing_layer(x), y))
+```
+
+While here, the pre-processing layer is the first in a larger model:
+
+```{r}
+input <- layer_input(shape = input_shape)
+output <- input %>%
+  preprocessing_layer() %>%
+  rest_of_the_model()
+model <- keras_model(input, output)
+```
+
+We'll talk about what way is preferable when, as well as showcase a few specialized layers, in a future post. Until then, please feel free to consult the -- detailed and example-rich - [vignette](https://github.com/rstudio/keras/blob/master/vignettes/new-guides/preprocessing_layers.Rmd).
+
+### Subclassing Python
+
+Imagine you wanted to port a Python model that made use of the following constraint:
+
+```{python}
+class NonNegative(tf.keras.constraints.Constraint):
+    def __call__(self, w):
+        return w * tf.cast(tf.math.greater_equal(w, 0.), w.dtype)
+```
+
+How can we have such a thing in R? Previously, there used to exist[^4] various methods to create Python-based objects, both R6-based and functional-style. The former, in all but the most straightforward cases, could be effort-rich and error-prone; the latter, elegant-in-style but hard to adapt to more advanced requirements.
+
+[^4]: We're not implying they don't exist anymore; they do. But the preferred method is now `%py_class%`.
+
+The new way, `%py_class%`, now allows for translating the above code like this:
+
+```{r}
+NonNegative(keras$constraints$Constraint) %py_class% {
+  "__call__" <- function(x) {
+    w * k_cast(w >= 0, k_floatx())
+  }
+}
+```
+
+Using `%py_class%`, we directly subclass the *Python* `object tf.keras.constraints.Constraint`, and override its `__call__` method.
+
+Why is this so powerful? The first advantage is clearly visible from the example: Translating Python code becomes a mechanical, almost, task. But there's more: The above method is independent from what *kind* of object you're subclassing. Want to implement a new layer? A callback? A loss? An optimizer? The procedure is always the same. No need to find a pre-defined R6 object in the `keras` codebase; one `%py_class%` delivers them all.
```

Comment:
```suggestion
Why is this so powerful? The first advantage is visible from the example: Translating Python code becomes an almost mechanical task. But there's more: The above method is independent from what *kind* of object you're subclassing. Want to implement a new layer? A callback? A loss? An optimizer? The procedure is always the same. No need to find a pre-defined R6 object in the `keras` codebase; one `%py_class%` delivers them all.
```

Reply:
Cool!

Reply:
streamlining my poetic commas away ;-))) ok ...

Reply:
I liked this comma here.